.github/workflows/ci-matrix-gen.yml

```diff
@@ -0,0 +1,297 @@
+name: Generate CI Matrix from ci.jsonnet
+
+on:
+  workflow_dispatch:
+    inputs:
+      jobs_to_run:
+        type: string
+        description: Job selection (Python regex)
+        required: false
+  workflow_call:
+    inputs:
+      jobs_to_run:
+        required: true
+        type: string
+      logs_retention_days:
+        required: false
+        type: number
+        default: 15
+      artifacts_retention_days:
+        required: false
+        type: number
+        default: 7
+
+jobs:
+  generate-tier1:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    env:
+      TARGET: tier1
+      JOBS: ${{ inputs.jobs_to_run }}
+    steps: &generate_matrix
+      - uses: actions/checkout@v4
+      - name: Download sjsonnet
+        run: |
+          curl -L -o sjsonnet https://github.com/databricks/sjsonnet/releases/download/0.5.7/sjsonnet-0.5.7-linux-x86_64
+          chmod +x sjsonnet
+      - name: Extract job matrix
+        id: set-matrix
+        run: |
+          python3 .github/scripts/extract_matrix.py ./sjsonnet ./ci.jsonnet ${TARGET} ${JOBS} > matrix.json
+          cat matrix.json
+          echo "matrix=$(cat matrix.json)" >> $GITHUB_OUTPUT
+
+  generate-tier2:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    env:
+      TARGET: tier2
+      JOBS: ${{ inputs.jobs_to_run }}
+    steps: *generate_matrix
+
+  generate-tier3:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    env:
+      TARGET: tier3
+      JOBS: ${{ inputs.jobs_to_run }}
+    steps: *generate_matrix
+
+  tier1:
+    needs: generate-tier1
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include: ${{ fromJson(needs.generate-tier1.outputs.matrix) }}
+    steps: &buildsteps
+
+      - name: Setup env
+        shell: bash
+        if: runner.os != 'Windows'
+        run: |
+          matrix_env='${{ toJson(matrix.env) }}'
+          echo $matrix_env | jq -r 'to_entries[] | "\(.key)=\(.value)"' >> tmp.env
+          while IFS= read -r line; do
+            eval "echo \"$line\"" >> $GITHUB_ENV
+          done < tmp.env
+          rm tmp.env
+
+      - name: Setup env (Windows)
+        if: runner.os == 'Windows'
+        shell: pwsh
+        run: |
+          $matrix_env = '${{ toJson(matrix.env) }}'
+          $envs = $matrix_env | ConvertFrom-Json
+          foreach ($pair in $envs.PSObject.Properties) {
+            $full_value = $pair.Value -replace '%([A-Za-z0-9_]+)%', '$env:$1'
+            $value = $ExecutionContext.InvokeCommand.ExpandString($full_value)
+            "$($pair.Name)=$value" | Out-File -FilePath $env:GITHUB_ENV -Append -Encoding utf8
+          }
+
+      - name: Actions/Checkout
+        uses: actions/checkout@main
```

Comment:
is `checkout@main` intentional, elsewhere it's `checkout@v4`. Pinned version is probably better?